usatoday.com: Rotating from landscape to portrait causes flickering

jesxilin opened the pull request:
#### 7eac12ec2ff9679924ce2d6adc86c943cacdccf4
<pre>
usatoday.com: Rotating from landscape to portrait causes flickering
<a href="https://bugs.webkit.org/show_bug.cgi?id=286367">https://bugs.webkit.org/show_bug.cgi?id=286367</a>
<a href="https://rdar.apple.com/93767145">rdar://93767145</a>

Reviewed by Wenson Hsieh.

usatoday.com has a visual viewport flickering issue when rotating from landscape to portrait.
This issue happens when rotating from portrait to landscape as well, but does not flicker due
to the width cutoffs set in javascript for the website.
Additionally, this issue actually happens when we change the viewport width in the meta tag.

In this scenario, the new width is calculated based on the current visual viewport scale and width.

In our code, the issue was as follows:
1. Document::processViewport is called whenever the viewport meta tag changes
2. The scale is set
    - Layout change happens before VisibleContentRectUpdates
3. VisualViewport::update calls setNeedsVisualViewportResize()
4. A resize is dispatched
    - This is the extra/premature resize
5. (UI) WebpageProxy::updateVisibleContentRects is called which in turn calls
    (WEB) WebPage::updateVisibleContentRects
6. A resize is dispatched
    - This is the correct resize that we anticipate with updated width

The scale is updated in the UI process but the width is updated in the Web process.

Thus, the scale is updated ahead of the width and hits this section of code in VisualViewport.cpp:
if (m_width != width || m_height != height || m_scale != scale) {
        if (document)
            document-&gt;setNeedsVisualViewportResize();

We hit m_scale != scale and call document-&gt;setNeedsVisualViewportResize(); prematurely,
when we should be waiting until either the width or height changes as well.

Thus, the issue of the extra resize(s) is fixed by removing the condition to check
for scale changes when determining when to call resize. This is safe because visual viewport
scales should never update independently from width/height.

An API test was also added for this change.

* Source/WebCore/page/VisualViewport.cpp:
(WebCore::VisualViewport::update):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/VisualViewport.mm: Added.
(TestWebKitAPI::TEST(VisualViewport, RotationResize)):

Canonical link: <a href="https://commits.webkit.org/289334@main">https://commits.webkit.org/289334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ffe2f3ce562406fb99370e9a28f57523067f5223

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86501 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91374 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37262 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88551 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6282 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14073 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66935 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24717 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89504 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78325 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47255 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4569 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32653 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36385 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93228 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13670 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9909 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75732 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74915 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18439 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19183 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17577 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13695 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/18963 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13443 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5426,6 +5426,21 @@ void Page::flushDeferredResizeEvents()
+    if (!m_shouldDeferSetNeedsVisualViewportResizeEvent)
+        return;
+    m_shouldDeferSetNeedsVisualViewportResizeEvent = false;
+    forEachDocument([&] (Document& document) {
```
Style guide nit: No space between the capture list and the argument list.

https://webkit.org/code-style-guidelines/#spacing-lambda-paren

```suggestion
    forEachDocument([&](Document& document) {
```

aprotyas replied:
Also, we don't need to capture anything here.
```suggestion
    forEachDocument([](Document& document) {
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4289,6 +4289,7 @@ static void handleAnimationActions(Element& element, uint32_t action)
     if (!localMainFrame)
         return;
 
+    m_page->startDeferringSetNeedsVisualViewportResizeEvent();
```
maybe this can be a SetForScope-like type of thing via a RAII wrapper? Probably not necessary though.

aprotyas replied:
`SetNeedsVisualViewportResizeEventDeferrer` 🤖

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,80 @@
+#if PLATFORM(IOS_FAMILY)
+@interface WKWebView ()
+- (void)_beginAnimatedResizeWithUpdates;
+@end
```
This is unnecessary.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,80 @@
+        [webView setFrame:CGRectMake(0, 0, [webView frame].size.height, [webView frame].size.width)];
+    }];
+
+    TestWebKitAPI::Util::runFor(3_s);
```
Should make this deterministic using `waitForNextPresentationUpdate`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,80 @@
+                var b = Math.round(a.width * a.scale);
+                b = 300 >= b ? 300 : 700 <= b ? 700 : "device-width";
+                c = document.querySelector("meta[name=viewport]").content = "width=" + b;
+                window.resizeCount = window.resizeCount + 1;
```
It would also be good to assert the values of `width` / `height` / `scale` inside the resize listener.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+#import <WebKit/WKWebView.h>
+#import <WebKit/WKWebViewConfiguration.h>
+#import <WebKit/WKWebViewPrivate.h>
+#import <WebKit/WKWebViewPrivateForTesting.h>
```
I don't think we're using anything from this header?

jesxilin replied:
Will take out the unnecessary headers!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+#import <WebKit/WKWebViewPrivateForTesting.h>
+#import <wtf/RetainPtr.h>
+
+namespace TestWebKitAPI {
```
```suggestion
namespace TestWebKitAPI {

```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+TEST(VisualViewport, RotationResize)
+{
+    RetainPtr configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
+    [configuration preferences]._logsPageMessagesToSystemConsoleEnabled = YES;
```
Delete, since this was for debugging only.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+{
+    RetainPtr configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
+    [configuration preferences]._logsPageMessagesToSystemConsoleEnabled = YES;
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 430, 812) configuration:configuration.get()]);
```
```suggestion
    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 430, 812) configuration:configuration.get()]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
```
```suggestion
#import "config.h"

#if PLATFORM(IOS_FAMILY)
```

pxlcoder replied:
And move the test into `/ios` rather than `/WebKitCocoa`

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/VisualViewport.mm:
```diff
@@ -0,0 +1,82 @@
+
+    EXPECT_EQ(windowResizeCount, 2);
+}
+}
```
```suggestion
}

#endif // PLATFORM(IOS_FAMILY)
```

lilyspiniolas commented on Tools/TestWebKitAPI/Tests/ios/VisualViewport.mm:
```diff
@@ -0,0 +1,81 @@
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 430, 812) configuration:configuration.get()]);
+
+    [webView synchronouslyLoadHTMLString:
+        @"<!DOCTYPE html>"
```
nit: using a raw string literal here instead of a multiline objective-c string could improve readability

jesxilin committed:
usatoday.com: Rotating from landscape to portrait causes flickering
https://bugs.webkit.org/show_bug.cgi?id=286367
rdar://93767145

Reviewed by Wenson Hsieh.

usatoday.com has a visual viewport flickering issue when rotating from landscape to portrait.
This issue happens when rotating from portrait to landscape as well, but does not flicker due
to the width cutoffs set in javascript for the website.
Additionally, this issue actually happens when we change the viewport width in the meta tag.

In this scenario, the new width is calculated based on the current visual viewport scale and width.

In our code, the issue was as follows:
1. Document::processViewport is called whenever the viewport meta tag changes
2. The scale is set
    - Layout change happens before VisibleContentRectUpdates
3. VisualViewport::update calls setNeedsVisualViewportResize()
4. A resize is dispatched
    - This is the extra/premature resize
5. (UI) WebpageProxy::updateVisibleContentRects is called which in turn calls
    (WEB) WebPage::updateVisibleContentRects
6. A resize is dispatched
    - This is the correct resize that we anticipate with updated width

The scale is updated in the UI process but the width is updated in the Web process.

Thus, the scale is updated ahead of the width and hits this section of code in VisualViewport.cpp:
if (m_width != width || m_height != height || m_scale != scale) {
        if (document)
            document->setNeedsVisualViewportResize();

We hit m_scale != scale and call document->setNeedsVisualViewportResize(); prematurely,
when we should be waiting until either the width or height changes as well.

Thus, the issue of the extra resize(s) is fixed by removing the condition to check
for scale changes when determining when to call resize. This is safe because visual viewport
scales should never update independently from width/height.

An API test was also added for this change.

* Source/WebCore/page/VisualViewport.cpp:
(WebCore::VisualViewport::update):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/VisualViewport.mm: Added.
(TestWebKitAPI::TEST(VisualViewport, RotationResize)):

Canonical link: https://commits.webkit.org/289334@main
